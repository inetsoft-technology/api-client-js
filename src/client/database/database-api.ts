/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * DatasourceProperties contains the properties of the configuration database connection.
 */
export interface DatabaseProperties {
   /**
    * The type of database. The allowed values are 'DB2', 'DERBY', 'H2', 'MYSQL', 'ORACLE',
    * 'POSTGRESQL', and 'SQL_SERVER'.
    */
   type: string;

   /**
    * The fully-qualified class name of the JDBC driver.
    */
   driver: string;

   /**
    * The JDBC URL of the database.
    */
   url: string;

   /**
    * The name of the default database or catalog. If not specified, the default database for the
    * login will be used.
    */
   defaultDatabase?: string;

   /**
    * The transaction isolation level for the database. The allowed values are 'READ UNCOMMITTED',
    * 'READ COMMITTED', 'REPEATABLE READ', and 'SERIALIZABLE'.
    */
   isolationLevel: string;

   /**
    * A flag that indicates if the database requires authentication.
    */
   requireLogin: boolean;

   /**
    * The user name used to authenticate with the database.
    */
   username?: string;

   /**
    * The password used to authenticate with the database.
    */
   password?: string;

   /**
    * The connection status of the database.
    */
   status?: string;
}

/**
 * DatabaseApi provides operations used to manage the configuration database.
 */
export interface DatabaseApi {
   /**
    * Gets the configuration database settings.
    *
    * @return the database settings.
    */
   getDatabase(): Promise<DatabaseProperties>;

   /**
    * Sets the configuration database settings.
    *
    * @param properties the database settings.
    *
    * @return the updated database settings.
    */
   setDatabase(properties: DatabaseProperties): Promise<DatabaseProperties>;

   /**
    * Installs the database schema. It is typically unnecessary to call this, as it is done
    * automatically.
    */
   installSchema(): Promise<void>;

   /**
    * Uninstalls the database schema. Extreme caution should be used when uninstalling the schema.
    * It causes all data to be irrevocably lost. It should never be done on a database that is
    * currently being used as a data space because it will make that server unusable.
    */
   uninstallSchema(): Promise<void>;
}
