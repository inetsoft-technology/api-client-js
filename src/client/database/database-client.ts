/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { AbstractClient } from "../abstract-client";
import { HttpAdapter } from "../http-adapter";
import { DatabaseApi, DatabaseProperties } from "./database-api";

/**
 * @hidden
 */
export class DatabaseClient extends AbstractClient implements DatabaseApi {

   constructor(adapter: HttpAdapter<any>, token: () => Promise<string>, baseUrl: string) {
      super(adapter, token, baseUrl);
   }

   public getDatabase(): Promise<DatabaseProperties> {
      return this.doGet("api/public/database");
   }

   public setDatabase(properties: DatabaseProperties): Promise<DatabaseProperties> {
      return this.doPut("api/public/database", properties);
   }

   public installSchema(): Promise<void> {
      return this.doPut("api/public/database/schema");
   }

   public uninstallSchema(): Promise<void> {
      return this.doDelete("api/public/database/schema");
   }
}
