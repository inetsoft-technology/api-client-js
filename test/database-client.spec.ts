/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { DatabaseClient } from "../src/client/database/database-client";

describe("DatabaseClient", () => {
   let adapter: any;
   let client: DatabaseClient;

   const database = {
      type: "MYSQL",
      driver: "com.mysql.jdbc.Driver",
      url: "jdbc:mysql://localhost:3306/inetsoft",
      isolationLevel: "READ COMMITTED",
      requireLogin: true,
      userName: "root",
      password: "password",
      status: "connected"
   };

   beforeEach(() => {
      adapter = {
         get: jest.fn(),
         post: jest.fn(),
         put: jest.fn(),
         delete: jest.fn(),
         map: jest.fn((promise: Promise<any>) => promise.then(response => response.data)),
         mapText: jest.fn(),
         mapBinary: jest.fn()
      };
      client = new DatabaseClient(adapter, () => Promise.resolve("secret"), "http://localhost:8080/sree/");
   });

   describe("when getDatabase() is called", () => {
      it("then the database properties are returned", () => {
         adapter.get.mockReturnValueOnce(Promise.resolve({data: database}));
         return client.getDatabase().then(result => {
            expect(result).toMatchSnapshot();
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/database");
         });
      });
   });

   describe("when setDatabase() is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.put.mockReturnValueOnce(Promise.resolve({data: database}));
         return client.setDatabase(database).then(result => {
            expect(result).toMatchSnapshot();
            expect(adapter.put.mock.calls.length).toBe(1);
            expect(adapter.put.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/database");
            expect(adapter.put.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when installSchema() is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.put.mockReturnValueOnce(Promise.resolve({data: {}}));
         return client.installSchema().then(() => {
            expect(adapter.put.mock.calls.length).toBe(1);
            expect(adapter.put.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/database/schema");
         });
      });
   });

   describe("when uninstallSchema() is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.delete.mockReturnValueOnce(Promise.resolve({data: {}}));
         return client.uninstallSchema().then(() => {
            expect(adapter.delete.mock.calls.length).toBe(1);
            expect(adapter.delete.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/database/schema");
         });
      });
   });
});
