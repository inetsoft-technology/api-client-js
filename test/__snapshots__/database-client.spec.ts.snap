// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatabaseClient when getDatabase() is called then the database properties are returned 1`] = `
Object {
  "driver": "com.mysql.jdbc.Driver",
  "isolationLevel": "READ COMMITTED",
  "password": "password",
  "requireLogin": true,
  "status": "connected",
  "type": "MYSQL",
  "url": "jdbc:mysql://localhost:3306/inetsoft",
  "userName": "root",
}
`;

exports[`DatabaseClient when setDatabase() is called then the adapter is called with the correct arguments 1`] = `
Object {
  "driver": "com.mysql.jdbc.Driver",
  "isolationLevel": "READ COMMITTED",
  "password": "password",
  "requireLogin": true,
  "status": "connected",
  "type": "MYSQL",
  "url": "jdbc:mysql://localhost:3306/inetsoft",
  "userName": "root",
}
`;

exports[`DatabaseClient when setDatabase() is called then the adapter is called with the correct arguments 2`] = `
Object {
  "driver": "com.mysql.jdbc.Driver",
  "isolationLevel": "READ COMMITTED",
  "password": "password",
  "requireLogin": true,
  "status": "connected",
  "type": "MYSQL",
  "url": "jdbc:mysql://localhost:3306/inetsoft",
  "userName": "root",
}
`;
